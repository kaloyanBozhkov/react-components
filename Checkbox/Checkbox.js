import React, { forwardRef, useState, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'

import styles from './styles.module.scss'

const setCheckedHandler = (checkbox, onChange, setChecked, isChecked) => {
    if (checkbox && checkbox.current) {
        // assume checkbox will change its check state, set it here so onChange func's checkbox.current has access to current state of checkbox, before applying changes to its state
        // basically gives onChange func access to what the next checked state will be, before choosing if to proceed with it or not
        checkbox.current.checked = !isChecked
    }

    // if set, run the onChange function which should return true if check state should be changed!
    // or  if there is no particular behavior on change of checkbox, just change the state of checked (animates check)
    if ((typeof onChange === 'function' && onChange({ target: checkbox.current })) || typeof onChange !== 'function') {
        
        setChecked(!isChecked)
    } else if (checkbox && checkbox.current) {
        // if parent has said that checkbox should not check/uncheck (based on return value of onChange), then reverse the checked value of html element

        checkbox.current.checked = isChecked
    }
}

const Checkbox = forwardRef(({ name, onChange = null, checked = false }, ref) => {
    const checkbox = useRef(null)
    const [isChecked, setChecked] = useState(checked)

    const classes = [
        styles.checkboxEl,
        isChecked ? styles.checked : ''
    ].join(' ').trim()

    // if a parent controls the checked state, update local state when passed checked prop changes
    useEffect(() => {
        setChecked(checked)
    }, [checked, onChange])

    // if a parent wants a ref to the input checkbox element
    useEffect(() => {
        // make sure ref is indeed passed down meaning we want parent to have it access to it
        if (ref) {
            ref.current = checkbox.current
        }
    }, [checkbox, ref])

    useEffect(() => {

        // if checkbox element's checked state does not match the state's checked state, then update it and trigger on Change.
        if (checkbox.current.checked !== isChecked) {
            checkbox.current.checked = isChecked
        }
    }, [isChecked])

    return (
        <div className={classes} onClick={() => setCheckedHandler(checkbox, onChange, setChecked, isChecked)}>
            <div />
            <input type='checkbox' ref={checkbox} name={name} />
        </div>
    )
})

Checkbox.propTypes = {
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    checked: PropTypes.bool
}

export default Checkbox