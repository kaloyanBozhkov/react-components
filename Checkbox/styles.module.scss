@import '../../../../scss/abstracts/variables';

.checkboxEl {
  > div {
    position: relative;
    &::before,
    &::after {
      content: '';
      background-color: $dark-grey;
      transition: 150ms;
      height: 0;
      width: 0;
    }
  }
  &.checked {
    > div {
      &::before {
        width: 30%;
        height: 2px;
        display: block;
        position: absolute;
        top: 60%;
        left: 20%;
        transform: rotate(35deg) translateY(-50%);
      }
      &::after {
        width: 2px;
        height: 50%;
        display: block;
        position: absolute;
        top: 25%;
        left: 60%;
        transform: rotate(35deg) translateX(-50%);
      }
    }
  }
  > div {
    background-color: white;
    border: 1px solid rgba(0, 60, 104, 0.1);
    width: 3.3rem;
    height: 3.3rem;
  }

  > input {
    display: none;
  }
}
