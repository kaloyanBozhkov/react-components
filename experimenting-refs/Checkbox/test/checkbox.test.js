import React, { createRef } from 'react'
import { shallow, mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

import Checkbox from '../Checkbox'

describe('testing checkbox atom', () => {
    
    it('should render the checkbox without a checked class', () => {
        const wrapper = shallow(<Checkbox name='someName' />)
        
        expect(wrapper).toMatchSnapshot()

        expect(wrapper.hasClass('checked')).toEqual(false)
    })

    it('should render checkbox checked by default', () => {
        const wrapper = shallow(<Checkbox name='someName' checked />)

        expect(wrapper.hasClass('checked')).toEqual(true)
    })

    it('should set ref to chekbox and have its checked to true by default', () => {
        const ref = createRef()
        const wrapper = mount(<Checkbox name='someName' ref={ref} checked />)

        expect(ref.current).toBeTruthy()
        expect(ref.current.checked).toEqual(true)
    })

    it('should handle custom on change event', () => {
        const mockFn = jest.fn(() => "some logic that parent wants to perform")
        const wrapper = shallow(<Checkbox name='someName' onChange={mockFn} />)

        wrapper.simulate('click')

        expect(mockFn).toHaveBeenCalled()
    })

    it('should handle custom on change event, and prevent check if return false', () => {
        // when the onChange returns false, stop checkbox from checking itself
        const onChange = jest.fn(() => false)
        const wrapper = shallow(<Checkbox name='someName' onChange={onChange} />)

        expect(wrapper.hasClass('checked')).toEqual(false)

        act(() => {
            wrapper.prop('onClick')()
        })
        
        expect(wrapper.hasClass('checked')).toEqual(false)
    })

    it('should handle custom on change event, and allow check if return true', () => {
        // when the onChange returns true, allow checkbox to check itself
        const onChange = jest.fn(() => true)
        const wrapper = shallow(<Checkbox name='someName' onChange={onChange} />)

        expect(wrapper.hasClass('checked')).toEqual(false)

        act(() => {
            wrapper.prop('onClick')()
        })

        expect(wrapper.hasClass('checked')).toEqual(true)
    })

    it('should keep html element refs check state unchanged, when check prevented by onChange', () => {
        const ref = createRef()
        const mockOnChange = jest.fn(() => false)
        const wrapper = mount(<Checkbox name='someName' ref={ref} onChange={mockOnChange} />)
        
        // get checkbox from within forward ref
        const checkbox = wrapper.childAt(0)

        expect(ref.current.checked).toEqual(false)
        
        act(() => {
            checkbox.prop('onClick')() 
        })
        
        expect(mockOnChange).toHaveBeenCalled()

        expect(ref.current.checked).toEqual(false)
    })

    it('should update html element refs check state, when check succeeds', () => {
        const ref = createRef()
        const mockOnChange = jest.fn(() => true)
        const wrapper = mount(<Checkbox name='someName' ref={ref} onChange={mockOnChange} />)
      
        // get checkbox from within forward ref
        const checkbox = wrapper.childAt(0)

        expect(ref.current.checked).toEqual(false)
        
        act(() => {
            checkbox.prop('onClick')() 
        })
        
        expect(mockOnChange).toHaveBeenCalled()

        expect(ref.current.checked).toEqual(true)
    })

    it('should toggle checked state based on clicks', () => {
        const wrapper = shallow(<Checkbox name='someName' />)

        expect(wrapper.hasClass('checked')).toEqual(false)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(true)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(false)
    })

    it('should toggle checked state based on clicks, with onChange preventing', () => {

        const mockOnChange = jest.fn(() => false)
        const wrapper = shallow(<Checkbox name='someName' onChange={mockOnChange} />)

        expect(wrapper.hasClass('checked')).toEqual(false)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(false)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(false)
    })

    it('should toggle checked state based on clicks, with onChange allowing', () => {

        const mockOnChange = jest.fn(() => true)
        const wrapper = shallow(<Checkbox name='someName' onChange={mockOnChange} />)

        expect(wrapper.hasClass('checked')).toEqual(false)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(true)

        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(false)
    })

    it('should give ref to html checkbox-type input element', () => {
        const inputRef = createRef()
        const wrapper = mount(<Checkbox name='someName' ref={inputRef} />)
        
        // test that current is not undefined, meaning it points to html elem: <input type='checkbox' />
        expect(inputRef.current).toBeTruthy()

        /// test that ref to html element allows us to get checked state
        expect(inputRef.current.checked).toEqual(false)

        wrapper.simulate('click')
        
        expect(inputRef.current.checked).toEqual(true)
    })

    it('should prevent checkbox from updating if parent says so through passed onChange func', () => {
        const inputRef = createRef()

        const wrapper = mount(<Checkbox name='someName' ref={inputRef} />)
        
        wrapper.simulate('click')

        expect(wrapper.hasClass('checked')).toEqual(false)

        expect(inputRef.current.checked).toEqual(true)
    })
})