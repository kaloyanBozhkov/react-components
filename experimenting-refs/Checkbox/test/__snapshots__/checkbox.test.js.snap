// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing checkbox atom should render the checkbox without a checked class 1`] = `
<div
  className="checkboxEl"
  onClick={[Function]}
>
  <div />
  <input
    name="someName"
    type="checkbox"
  />
</div>
`;
